allow use of cloudnetworks in ssh_interface

gtmanfred opened the pull request:
I wanted to add this in 2015.8, but was held up by an issue that appears to have been fixed in develop and 2015.8.
#24855

I am hoping I can get this allowed back into 2015.8.

Thanks!
Daniel

gtmanfred committed:
add ability to specify a cloudnetwork in ssh_interfaces

gtmanfred committed:
add a note in the documentation for openstack

about the cloud networks in ssh_interfaces

gtmanfred commented:
bah, i think this is broken

techhat commented on salt/cloud/clouds/nova.py:
```diff
@@ -86,7 +86,7 @@
 
 Note: For rackconnect v3, rackconnectv3 needs to be specified with the
-rackconnect v3 cloud network as it's variable
+rackconnect v3 cloud network as it's variable.
```
While we're fixing up this line, we might as well also change "it's" to "its".

techhat commented:
+!

gtmanfred commented:
yes

gtmanfred commented:
Actually didn't mean to fix that line :P

gtmanfred committed:
fix appostrophe

gtmanfred commented:
Ok, this is all good, but there looks to be a problem with map files.

The information in this gets set if you do it all from the command line, but doesn't with map files.

@techhat @rallytime I am not sure why that is happening, but it works if you build one on the command line, or build a bunch in parrallel, but it doesn't set cloudnetwork to True in vm_ if it is all built from a map file :/

Thanks
Daniel

gtmanfred commented:
works
```
salt-cloud -p centos-2-nova gtmanfred.com
```
```
salt-cloud -Py -p centos-2-nova pg{1..4}.gtmanfred.com master{1..2}.gtmanfred.com
```

doesn't work
```
salt-cloud -Pym map
```

```
[root@salt ~]# salt --versions-report
Salt Version:
           Salt: 2015.8.0rc1-49-g5185939

Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.6.0
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
           RAET: Not Installed
        Tornado: 4.1
            ZMQ: 3.2.5
          ioflo: Not Installed
        libnacl: 1.4.2
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
       pycrypto: 2.6.1

System Versions:
           dist: centos 7.1.1503 Core
        machine: x86_64
        release: 3.10.0-229.1.2.el7.x86_64
         system: CentOS Linux 7.1.1503 Core
```

And I am stripping the data from my cloud configs.

gtmanfred commented:
```
[root@salt salt]# tail -c +0 /etc/salt/cloud /etc/salt/cloud.profiles.d/nova.conf /etc/salt/cloud.providers.d/nova.conf
==> /etc/salt/cloud <==
update_cachedir: True
diff_cache_events: True
ssh_interface: <cloudnetwork>
change_password: True
wait_for_ip_timeout: 6000
display_ssh_output: False
protocol: ipv4
script_args: -LP git 2015.8
minion:
  master: <salt>
  #startup_states: highstate
  mysql.default_file: '/root/.my.cnf'
  mine_functions:
    network.ip_addrs: []
  tcp_keepalive: True
  tcp_keepalive_idle: 30

==> /etc/salt/cloud.profiles.d/nova.conf <==
centos-2-nova:
  provider: my-nova
  size: performance1-2
  image: CentOS 7 (PVHVM)
  ssh_key_name: saltstack
  flush_mine_on_destroy: True
  networks:
    - net-id: 11111111-1111-1111-1111-111111111111
    - net-id: 00000000-0000-0000-0000-000000000000
    - net-id: <cloudnetwork uuid>


==> /etc/salt/cloud.providers.d/nova.conf <==
my-nova:
  identity_url: 'https://identity.api.rackspacecloud.com/v2.0/'
  compute_region: <region>
  user: <username>
  tenant: <tenantid>
  provider: nova
  api_key: <apikey>
  os_auth_plugin: rackspace
```

gtmanfred commented:
I have an idea that might fix this problem.

gtmanfred commented:
for some reason salt_cloud does not see /etc/salt/cloud when running in map mode.

Or it doesn't want to use options set there, I see the ssh_interface set to saltstack in __opts__ when running with a map, but it doesn't want to actually work unless I flip the search to use global

gtmanfred commented:
yup, as soon as I move it to the provider, everything works

gtmanfred committed:
make sure to set access_ip

Note: ssh_interface has to be in the provider for this to work with map
files, right now at least

rallytime commented:
Good find @gtmanfred! Do you have profiles set in both `/etc/salt/cloud` and `/etc/salt/cloud.profiles*`? I don't think you're supposed to mix those...I am fairly certain we are not supporting mixing those two configurations, which might explain why that config wasn't getting picked up.

gtmanfred commented:
I only have profiles in that one profiles.d file.

The config I have above is everything I have setup currently, with the exception of ssh_interface now being in the provider.

It is really wierd, it works in parallel, if you put all the minions on the commandline, but with a map file, it doesn't seem to see the ssh_interface in /etc/salt/cloud

rallytime commented:
That is really weird. There must be a bug in there somewhere when we're loading in the map file...

gtmanfred commented:
Actually, I appear to be mistaken, I just checked on 2015.5 at least and it doesn't appear to accept ssh_interface: private_ips at all :/

I could have sworn it worked in the past.

Oh well, good to know that it can't be in /etc/salt/cloud now I guess.

Once this finshes it's checks, it should be all set to go.

rallytime commented:
Looks good! Thanks!

gtmanfred commented:
@rallytime how often do yall do forward merges?

Thanks
Daniel

rallytime commented:
@gtmanfred Every couple of days. @basepi is the one that usually does them. I'm sure he'll do one soon. :)
